Part 4: Multi db improvements, Basic API for connection switching

eileencodes opened the pull request:
This PR implements the basic API requirements laid out in https://github.com/rails/rails/pull/33877 by DHH. The PR aims to focus only on implementing the `connects_to` and `connected_to` API. For now it does not tackle any configuration changes (we can hash that out in future PRs). If this API is acceptable I will add tests.

cc/ @dhh @matthewd @rafaelfranca @tenderlove 

---

This PR adds the ability to 1) connect to multiple databases in a model,
and 2) switch between those connections using a block.

To connect a model to a set of databases for writing and reading use
the following API. This API supersedes `establish_connection`. The
`writing` and `reading` keys represent handler / mode names and
`animals` and `animals_replica` represents the database key to look up
the configuration hash from.

```ruby
class AnimalsBase < ApplicationRecord
  connects_to database: { writing: :animals, reading: :animals_replica }
end
```

Inside the application - outside the model declaration - we can switch
connections with a block call to `connected_to`.

If we want to connect to a db that isn't default (ie readonly_slow) we
can connect like this:

Outside the model we may want to connect to a new database (one that is
not in the default writing/reading set) - for example a slow replica for
making slow queries. To do this we have the `connected_to` method that
takes a `database` hash that matches the signature of `connects_to`. The
`connected_to` method also takes a block.

```ruby
ModelInPrimary.connected_to(database: { slow_readonly: :primary_replica_slow }) do
  ModelInPrimary.do_something_thats_slow
end
```

For models that are already loaded and connections that are already
connected, `connected_to` doesn't need to pass in a `database` because
you may want to run queries against multiple databases using a specific
mode/handler.

In this case `connected_to` can take a `handler` and use that to swap on
the connection passed. This simplies queries - and matches how we do it
in GitHub. Once you're connected to the database you don't need to
re-connect, we assume the connection is in the pool and simply pass the
handler we'd like to swap on.

```ruby
ActiveRecord::Base.connected_to(hander: :reading) do
  Dog.read_something_from_dog
  ModelInPrimary.do_something_from_model_in_primary
end
```

eileencodes committed:
Basic API for connection switching

This PR adds the ability to 1) connect to multiple databases in a model,
and 2) switch between those connections using a block.

To connect a model to a set of databases for writing and reading use
the following API. This API supercedes `establish_connection`. The
`writing` and `reading` keys represent handler / role names and
`animals` and `animals_replica` represents the database key to look up
the configuration hash from.

```
class AnimalsBase < ApplicationRecord
  connects_to database: { writing: :animals, reading: :animals_replica }
end
```

Inside the application - outside the model declaration - we can switch
connections with a block call to `connected_to`.

If we want to connect to a db that isn't default (ie readonly_slow) we
can connect like this:

Outside the model we may want to connect to a new database (one that is
not in the default writing/reading set) - for example a slow replica for
making slow queries. To do this we have the `connected_to` method that
takes a `database` hash that matches the signature of `connects_to`. The
`connected_to` method also takes a block.

```
AcitveRecord::Base.connected_to(database: { slow_readonly: :primary_replica_slow }) do
  ModelInPrimary.do_something_thats_slow
end
```

For models that are already loaded and connections that are already
connected, `connected_to` doesn't need to pass in a `database` because
you may want to run queries against multiple databases using a specific
role/handler.

In this case `connected_to` can take a `role` and use that to swap on
the connection passed. This simplies queries - and matches how we do it
in GitHub. Once you're connected to the database you don't need to
re-connect, we assume the connection is in the pool and simply pass the
handler we'd like to swap on.

```
ActiveRecord::Base.connected_to(role: :reading) do
  Dog.read_something_from_dog
  ModelInPrimary.do_something_from_model_in_primary
end
```

tenderlove commented on activerecord/lib/active_record/connection_handling.rb:
```diff
@@ -47,6 +47,95 @@ module ConnectionHandling
+    def connected_to(database: nil, handler: nil, &blk)
+      if database && handler
+        raise ArgumentError, "connected_to can only accept handler or database, but not both arguments."
+      end
```
Should we raise an exception if both are `nil`?

eileencodes commented:
> What would happen if two models have the same handler name for two different databases? Should we support that? Say:

That's exactly what I want to support because that's how we do it at GitHub. We have 10 connections that belong to a `writing` handler (we call it default) and 10 connections that belong to a `reading` handler (we call it readonly). I'm a little confused because I thought you said at Shopify you use the multiple handler approach as well? The underlying connection swapping behavior here isn't different from the original PR, just the public API is.

In GitHub we'd don't write this:

```ruby
Dog.connected_to(handler: :reading) do
  Dog.create! # explode from Dog bc doing a write on a read
  Book.create! # isn't called but not because it's Dog's handler, you told Rails what handler to use - `reading`. 
end
```

Instead we write this (but with GitHub instead of Ar Base bc Rails doesn't support this yet)

```ruby
ActiveRecord::Base.connected_to(handler: :reading) do
  Dog.create!
  Book.create!
end
```

If we want to write to multiple dbs we can do that by using the writing handler:

```ruby
ActiveRecord::Base.connected_to(handler: :writing) do
  Dog.create! # success
  Book.create! # success
end
```

`Dog` and `Book` know which database they belong to because the model tells them. The `connected_to` method looks up the handler, and then Rails looks up the connection from that handler with the connection specification name.

tenderlove commented:
@rafaelfranca I think we should support your first scenario, but if you want to switch both models you need to do it at `AR::Base` as @eileencodes mentions.  Maybe we should raise an exception if you don't call it on `AR::Base`?

dhh commented:
I like this. A few notes:

If you're connecting directly to a specific database, you shouldn't have to declare the role:

```ruby
ModelInPrimary.connected_to(database: :primary_replica_slow) do
  ModelInPrimary.do_something_thats_slow
end
```

Re: handler, I don't really like that word much. I'd prefer to use "role". That would connect with the future 3-tier database.yml configuration setup as well. So it would be:

```ruby
ActiveRecord::Base.connected_to(role: :reading) do
  Dog.create!
  Book.create! # Will raise if a :reading role isn't found on Book
end
```

@tenderlove I'd be curious to see how many instances of connection switching you have in the code? I was initially partial to having some syntatic sugar, but I don't think switching roles mid-flight is going to be a super common action. And if it isn't, then I'd rather be as clear as possible about what's going on.

On the larger topic of r/w splitting, @eileencodes, you're working towards a place where AR automatically will pick the :writing role when AR is doing INSERTs and :reading role when AR is doing SELECTs, right? I thought there was some confusion about whether that's within this initial scope of work when discussing with @matthewd in the earlier thread.

eileencodes commented:
> If you're connecting directly to a specific database, you shouldn't have to declare the role

👍 I will work on changing this requirement. Currently I have it so it creates a new `handler` (or role) to organize the connections. But that's not actually necessary now that I think about it so I'll adjust this PR accordingly.

> Re: handler, I don't really like that word much

I can change this. For background handler makes sense to me since it is switching on the connection_handler - but perhaps that's too much for the user to need to know. 

> On the larger topic of r/w splitting, @eileencodes, you're working towards a place where AR automatically will pick the :writing role when AR is doing INSERTs and :reading role when AR is doing SELECTs, right

Yes but this is further down the line (ie not for this PR). Rails needs to be able to switch connections before it can know what to switch to.

> I'd be curious to see how many instances of connection switching you have in the code?

We actually do this quite a bit since we default to the replicas, expect in certain circumstances where we need to explicitly call readonly. 

* 296 block calls to readonly db in app and lib
* 3 block calls to write db in lib
* 6 block calls to a dynamic switcher that chooses the db in app and lib.

deepj commented:
A question: Can this switching be used as a failover? Let's say, a primary connection failed, switch to secondary (backup) one.

eileencodes commented:
@dhh we default to read but switch on the request type (GET == read, POST == write) rather than the sql query. So in some cases we need to switch back to the read or to the write in order to handle that. I assume we will need less of those if we have Rails auto switch based on SQL rather than request type. I think that if we really do need the helper methods we can add those later.

@deepj No. We're quite a bit aways from something like that.

rafaelfranca commented:
> That's exactly what I want to support because that's how we do it at GitHub. We have 10 connections that belong to a writing handler (we call it default) and 10 connections that belong to a reading handler (we call it readonly). I'm a little confused because I thought you said at Shopify you use the multiple handler approach as well? The underlying connection swapping behavior here isn't different from the original PR, just the public API is.

Yes, we use the multiple handlers, but it seems the implementation store the name of the handlers in `ActiveRecord::Base` so it means two models can't have the same handler name because the later will override the former definition even if you define in two different models. Am I getting the implementation wrong?

Would not:

```
ActiveRecord::Base.connected_to(handler: :reading) do
  Dog.create!
  Book.create!
end
```

fail because `Dog` will be connected to the same database than `Book` (`things`, not `animals` given `Book` was define after `Dog`)?

rafaelfranca commented:
Ok I think I get it. The handler is the same for all models but it holds a connection pool for each model with a different `connection_specification_name`.

I agree with DHH's suggestions for the API.

👍 from me.

eileencodes commented:
> Ok I think I get it. The handler is the same for all models but it holds a connection pool for each model with a different connection_specification_name.

Yup! That's exactly how it works.

I'm writing up some tests and will be pushing up later this weekend or early next week. I think we're almost ready to merge this (with DHH's changes). That will unblock a lot of the future work. 😄 

Also @matthewd originally had some concerns about threads but we paired today and found it's not a problem. The connection handler is thread local so we're good there 👍

eileencodes commented:
* Changed `handler` to `mode`
* Changed `connected_to` to take a single database instead of a hash since that's for connecting to a specific database / role. 
* Raise if both `mode` and `handler` are nil in `connected_to`
* Added tests
* Added docs (note I don't think we're ready to update the guides yet so I left those. I'd rather be able to tell a cohesive story at the end)

dhh commented:
I think `mode` is an improvement over `handler`, but I'm not sure it's quite enough. If you think about modes when opening files, it's the mode with which you open the same file that's designated. Not a different file. So the extension here is that `mode` would refer to different ways of connecting to the same database, not to different databases.

That's why I like `role`. That implies that there are multiple actors (databases) that play a specific role in the application. And role is also flexible enough to work with things like `:statistics` or `:analytics` where `mode` would be awkward.

eileencodes commented:
`mode` switched to `role` 👍

rafaelfranca commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

rafaelfranca commented:
Thank you so much @eileencodes for working on this.

dreyks commented:
cool! great features are coming!

what about stickiness? e.g. when a write occurs temporarily re-route reads to default db too

jeremy commented:
@dreyks Please do start building it! 😍 This is the underlying API needed to implement things like automatic read/write splitting and stickiness.

eileencodes commented:
@jeremy I'm already working on automatic switching.

rafbm commented:
Is this multi-DB support only for primary-replica setups? Curious to know if there are any plans for supporting migrations / rake tasks on different databases? In our case, we do all business logic using regular AR models connected to our main DB but we have a separate DB for logs (which don’t require ACID guarantees, can live on a DB with lower CPU / RAM specs, etc).

What we currently do is pretty simple: we keep a `/db_logs` structure identical to the `/db` one along with a `config/database_logs.yml` next to `config/database.yml`, which have the regular YAML structure (nothing custom).

<img width="238" alt="screen shot 2018-10-14 at 22 15 50" src="https://user-images.githubusercontent.com/166943/46926646-951b3b00-cfff-11e8-998e-06a12da7d360.png">

To support this, our only custom piece of code is a mapping of paths in `config/application.rb`:

```ruby
require_relative 'boot'

# ...

module Missive
  class Application < Rails::Application
    #
    # Support rake tasks on separate logs DB
    #
    if ENV['DB'] == 'logs'
      {
        'config/database' => 'config/database_logs.yml',
        'db'              => 'db_logs',
        'db/migrate'      => 'db_logs/migrate',
        'db/seeds.rb'     => 'db_logs/seeds.rb',
      }.each do |a, b|
        config.paths[a] = Rails.root.join(b)
      end
    end

    # ...
  end
end
```

This lets us run migrations on the logs DB by adding `DB=logs` on the command line:

```yaml
# main DB
bundle exec rake db:migrate:status

# logs DB
bundle exec rake db:migrate:status DB=logs
```

The following part isn’t relevant to the question but for completeness, we configure log models (for web requests and background workers) like so:

```ruby
# app/models/log_record.rb
class LogRecord < ApplicationRecord
  self.abstract_class = true

  establish_connection begin
    erb = File.read(Rails.root.join('config/database_logs.yml'))
    YAML.load(ERB.new(erb).result)[Rails.env]
  end
end

# app/models/request_log.rb
class RequestLog < LogRecord
  # ...
end

# app/models/worker_log.rb
class WorkerLog < LogRecord
  # ...
end
```

As you can see, it was pretty trivial for us to implement this pattern, yet the `config.paths` bit in `config/application.rb` does feel like a little hack and it would be nice to have an official way to deal with migrations affecting separate DBs.

rafbm commented:
Ah, silly me! I hadn’t thought of searching for other “Part X” pull requests from Eileen. Answer to my question is #32274. I don’t believe the preliminary [changelog](https://github.com/eileencodes/rails/blob/31021a8c85bb80300deb01db96876858ff417f4a/activerecord/CHANGELOG.md) mentions these rake tasks; could be a good thing to add.

salimane commented:
It seems that joins are not being correctly handled``joins`` across databases

```

class Person < ApplicationRecord
  has_one :animal,
          class_name: 'Animal',
          foreign_key: 'person_id',
          dependent: :nullify, inverse_of: :person
end

class Animal < ApplicationRecord
  self.abstract_class = true
  connects_to database: { reading: :animals, writing: :animals }
  belongs_to :person, foreign_key: 'person_id', inverse_of: :animal
end

Person.joins(:animal).to_sql
=> "SELECT `persons`.* FROM `persons` INNER JOIN `animal` ON `animal`.`person_id` = `persons`.`person_id`"
```

or am I missing something @eileencodes ?

eileencodes commented:
No you're not missing something, Rails does not yet handle joining across separate databases. We're working on supporting the ability for Rails to recognize the connections are different and to split up the queries into 2 selects but the join syntax isn't going to be possible across 2 machines.

salimane commented:
Can I suggest something: during ``joins`` and related, check if the 2 classes share the same host.
-  if yes, add the database if DB names are different.
- if no, raise an error suggesting that the ``joins`` is impossible

what do you think @eileencodes ?

voordev commented:
Perhaps a stupid wuestion but does this enable:  

1. Multitennant setup where organisation entity loads its own database
2. Organisation.organisation_customers.all would return only records from current organisation database?
3. This would solve the " select users from organisation where organisation_id = x " queries and enable " select users from organisation "?

Is there a way to add and or create databases at runtime? From what i understand now you have to append db config to databasy.yml

I could not find a conclusive post on this.
<3 from Amsterdam to rails core team and all contributors to make this happen.

voordev commented:
Nobody?

eileencodes commented:
No questions are stupid @voordev, however I'm not 100% I understand what you're asking. I've started working on documentation here https://github.com/rails/rails/pull/36389 that will hopefully answer some of your questions about how Rails works.

voordev commented:
@eileencodes

I mean does rails 6 support multi tennant. Entity universities with 1 database per entity. or can we better use other not to be named solutions?

eileencodes commented:
As the docs note, no, not yet, Rails doesn't support sharding.

vitobotta commented:
Hi! May I ask what is the current status? I would like to be able to safely split reads/writes between master and slaves in a MySQL replication, automatically. Is this possible yet? Thanks in advance.

mujz commented:
Awesome work! 
There's a typo in the PR's description. `hander` should be `handler`.

concept47 commented:
I had one question. Is it possible to specify a series of replicas in database.yml for reading so that when you do something like this

    ActiveRecord::Base.connected_to(handler: :reading) do
      Dog.read_something_from_dog
      ModelInPrimary.do_something_from_model_in_primary
    end

it can go to one of the different dbs you have specified in that group ... vs just one specific db?

shekhar098 commented:
ActiveRecord::Base.connected_to(database: :key_logs) do . I want manual db switching. connected_to method is looking for adaptor details in database.yml. 

I have multiple database with same details. I don't want multiple schema . Rails 6 is working on schema switching?

deivinsontejeda commented:
Hi @eileencodes,

I'm working on this features. let me share quickly my use case... 
- We are working on a multi-tenant, but using different physically DB (each customer a DB).
- The DB is create when a new customer is registered. (We have some process which will create the DB)
- When DB was created, it does not exist in `connects_to`. Here is my question, we can run `ApplicationRecord.connects_to` once app is booted in order to register connection of new DB?

Leave some code to help you see my current scenario...

database.yml
```yml
test:
  shard:
    <<: *default
    url: <%= ENV.fetch('DATABASE_URL_SHARD_TEST') %>
    migrations_paths: db/shard_migrate

  shard_replica:
    <<: *default
    url: <%= ENV.fetch('DATABASE_URL_SHARD_TEST') %>
    replica: true

  global:
    <<: *default
    url: <%= ENV.fetch('DATABASE_URL_GLOBAL_TEST') %>
    migrations_paths: db/global_migrate

  global_replica:
    <<: *default
    url: <%= ENV.fetch('DATABASE_URL_GLOBAL_TEST') %>
    replica: true
```

Then, `ApplicationRecord`

```ruby
class ApplicationRecord < ActiveRecord::Base
  self.abstract_class = true

  # We have some code to read current config and return the Hash with shards available in database.yml
  connects_to shards: { global: { writing: :global, reading: :global_replica }, shard: { writing: :shard, reading: :shard_replica } }
  scope :active, -> { where(active: true) }
end
```

Until here no mayor issues, I can switch connections and everything looks ok... 

But now, imagine I created a new DB connection which is not part of `database.yml` (I have some internal process to read them).

Again my question, I would like to reload `ApplicationRecord.connects_to` in order to register that new connection... 

```ruby
ApplicationRecord.connects_to shards: { global: { writing: :global, reading: :global_replica }, shard: { writing: :shard, reading: :shard_replica }, customer1: { writting: :customer1, reading: :customer1_replica} }
```

The `customer1` is dynamically fetched/added. In my local tests when I do that, I got `ActiveRecord::ConnectionNotEstablished`.

```ruby
ActiveRecord::Base.connected_to(shard: :company1) do
  Model.find(id)
end
```

Any thoughts? 

I appreciate your feedback :+1:

eileencodes commented:
> Again my question, I would like to reload ApplicationRecord.connects_to in order to register that new connection...

No this isn't allowed - doing this would clobber all existing connections to add the new one and potentially in the middle of a request. That's super dangerous so even if there was a workaround I wouldn't recommend it. There's also other potential issues with doing a setup like this and it's safest to reload the application when messing with database connections.

The recommended way of doing this (or at least the way we discussed at my prior job of working around this) is to pre-setup your shards. For example, say you currently have 4 customers. Instead of only setting up 4 shards for each existing customer you'd set up 100 shards and add the connections for those. Then in your global router table all you would need to do is insert a new record for the new tenant when they sign up. The connections will be active but inaccessible until the customer is added to the DB. 

The other way is add a new config and new connection and then deploy the app each time you need to add a new customer.

---

In the future it's better to open a new issue or ask questions on the forum. I occasionally unsubscribe from old PRs and it also hides these questions from future readers looking for the same answer
